MoreLinq/Partition.cs

```diff
@@ -0,0 +1,297 @@
+        /// Partitions a grouping by nullable Boolean keys into a projection of
+        /// true elements, false elements and null elements, respectively.
+        /// </summary>
+        /// <typeparam name="T">Type of elements in source groupings.</typeparam>
+        /// <typeparam name="TResult">Type of the result.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="resultSelector">
+        /// Function that projects the result from sequences of true elements,
+        /// false elements and null elements, respectively, passed as
+        /// arguments.
+        /// </param>
+        /// <returns>
+        /// The return value from <paramref name="resultSelector"/>.
+        /// </returns>
+
+        public static TResult Partition<T, TResult>(this IEnumerable<IGrouping<bool?, T>> source,
+            Func<IEnumerable<T>, IEnumerable<T>, IEnumerable<T>, TResult> resultSelector)
+        {
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+            return source.Partition(true, false, null, (t, f, n, _) => resultSelector(t, f, n));
+        }
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching a key
+        /// and those groups that do not.
+        /// </summary>
+        /// <typeparam name="TKey">Type of keys in source groupings.</typeparam>
+        /// <typeparam name="TElement">Type of elements in source groupings.</typeparam>
+        /// <typeparam name="TResult">Type of the result.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="key">The key to partition.</param>
+        /// <param name="resultSelector">
+        /// Function that projects the result from sequences of elements
+        /// matching <paramref name="key"/> and those groups that do not,
+        /// passed as arguments.
+        /// </param>
+        /// <returns>
+        /// The return value from <paramref name="resultSelector"/>.
+        /// </returns>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey key,
+            Func<IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            Partition(source, key, null, resultSelector);
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching a key
+        /// and those groups that do not. An additional parameter specifies how
+        /// to compare keys for equality.
+        /// </summary>
+        /// <typeparam name="TKey">Type of keys in source groupings.</typeparam>
+        /// <typeparam name="TElement">Type of elements in source groupings.</typeparam>
+        /// <typeparam name="TResult">Type of the result.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="key">The key to partition.</param>
```

Comment:
This may be my non-native English, but I fail to parse this description. I think it would be clearer to say: `The key of the matching partition`. Mutatis mutandis, the same applies to the overloads with more `key` arguments.